### PR TITLE
Validations/unique operation names

### PR DESCRIPTION
**Pull request by freshtonic**
I've created a validator that works with the AST visitor/reducer infrastructure.

This visitor is so simple that it does not need anything from the TypeInfoVisitor to do its job.

I'm seeking feedback for anything and everything (overall approach, stylistic stuff etc). Don't hold back!

ParallelVisitor still needs tests (it's tested indirectly right now).

Next I'd like to port `FieldsOnCorrectType` because that will exercise the `TypeInfoVisitor`.

### CODE REVIEW

**Comment by joshprice** [comment]
Will review after merging #57 and rebasing to see diff

File: lib/graphql/lang/ast/type_info.ex [hunk 1]
```diff
@@ -58,7 +58,12 @@ defmodule GraphQL.Lang.AST.TypeInfo do
       name == Introspection.meta(:typename)[:name] ->
         Introspection.meta(:typename)
       parent_type.__struct__ == GraphQL.Type.ObjectType || parent_type.__struct__ == GraphQL.Type.Interface ->
-        parent_type.fields[name]
+        # FIXME: this "function or map" logic is repeated in the executor. DRY IT UP.
```
Comment (freshtonic) [inline lib/graphql/lang/ast/type_info.ex:61]: This logic is repeated in two places. It seems like it should be on one of the GraphQL types (or a `Fields` protocol that's implemented by `ObjectType`, `Interface` and `Union`?)

@joshprice @markolson what do you think?

**Comment by freshtonic** [comment]
There is remaining work on this validator: I need to bring over all of the other tests for this validation rule form graphql-js. However, if I can merge it early it means other people have an example of how to implement validators and we can start working in parallel. @joshprice fine to merge?

**Comment by freshtonic** [comment]
So I implemented all the other tests for this rule (from graphql-js).

File: lib/graphql/lang/ast/parallel_visitor.ex [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+          accumulator[:skipping][child_visitor] == node ->
+            put_in(acc[:skipping][child_visitor], false)
+          true -> accumulator
+        end
+      end)
+      { :continue, accumulator }
+    end
+
+    defp skipping?(accumulator, child_visitor) do
+      !!accumulator[:skipping][child_visitor]
```
Comment (joshprice) [inline lib/graphql/lang/ast/parallel_visitor.ex]: :boom: :boom: :smile:

File: lib/graphql/validation/rules/unique_operation_names.ex [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    end
+  end
+
+  defimpl Visitor do
+    def enter(_visitor, node, accumulator) do
+      if node.kind == :OperationDefinition do
+        if Map.has_key?(node, :name) do
+          op_name = node.name
+          if op_name.value do
+            if accumulator[:operation_names][op_name.value] do
```
Comment (joshprice) [inline lib/graphql/validation/rules/unique_operation_names.ex]: 4 nested `if`s need some extraction

**Comment by joshprice** [comment]
:rocket:

**Pull request merged**